### PR TITLE
[feature] partition prune support of list partition

### PR DESCRIPTION
**Pull request by hiliuxg**
issue #4035

## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Problem Summary(Required) ：
olap table partition prune support of list partition

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 90 / 99 (90.91%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PartitionPruneRule.java|7|10|70.00%|[88, 89, 91]|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|7|10|70.00%|[252, 253, 255]|
|:large_blue_circle:|com/starrocks/planner/ListPartitionForOlapPruner.java|76|79|96.20%|[36, 47, 70]|

**Comment by wyb** [comment]
I think you can use the current `ListPartitionPruner` directly.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PartitionPruneRule.java [hunk 1]
```diff
@@ -72,6 +85,69 @@ public List<OptExpression> transform(OptExpression input, OptimizerContext conte
                 input.getInputs()));
     }
 
+    private void putValueMapItem(TreeMap<LiteralExpr, Set<Long>> columnToPartitionValuesMap,
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PartitionPruneRule.java:88]:
```suggestion
    private void putValueMapItem(TreeMap<LiteralExpr, Set<Long>> partitionValueToIds,
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PartitionPruneRule.java [hunk 2]
```diff
@@ -72,6 +85,69 @@ public List<OptExpression> transform(OptExpression input, OptimizerContext conte
...
+    private List<Long> listPartitionPrune(ListPartitionInfo listPartitionInfo,
+                                          LogicalOlapScanOperator olapScanOperator) {
+
+        Map<ColumnRefOperator, TreeMap<LiteralExpr, Set<Long>>> columnToPartitionValuesMap = new HashMap<>();
+        Map<ColumnRefOperator, Set<Long>> columnToNullPartitions = new HashMap<>();
+
+        // single item list partition has only one column mapper
+        Map<Long, List<LiteralExpr>> literalExprValuesMap = listPartitionInfo.getLiteralExprValues();
+        if (literalExprValuesMap != null && literalExprValuesMap.size() > 0) {
+            TreeMap<LiteralExpr, Set<Long>> columnToPartitionValuesItem = new TreeMap<>();
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PartitionPruneRule.java:108]:
```suggestion
            TreeMap<LiteralExpr, Set<Long>> partitionValueToIds = new TreeMap<>();
```

File: fe/fe-core/src/test/java/com/starrocks/sql/optimizer/rule/transformation/PartitionPruneRuleTest.java [hunk 3]
```diff
@@ -32,13 +36,15 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNull;
 
-public class PartitionPruneRuleTest {
+public class PartitionPruneRuleTest{
```
Comment (Seaven) [inline fe/fe-core/src/test/java/com/starrocks/sql/optimizer/rule/transformation/PartitionPruneRuleTest.java:47]:
```suggestion
public class PartitionPruneRuleTest {
```

**Comment by Seaven** [comment]
reslove conflicts plz~

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PartitionPruneRule.java [hunk 4]
```diff
@@ -72,6 +85,69 @@ public List<OptExpression> transform(OptExpression input, OptimizerContext conte
...
+                columnToNullPartitions, scalarOperatorList);
+        try {
+            return partitionPruner.prune();
+        } catch (AnalysisException e) {
+            LOG.warn("PartitionPrune Failed. ", e);
+        }
+        return null;
+    }
+
     private List<Long> partitionPrune(OlapTable olapTable, RangePartitionInfo partitionInfo,
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PartitionPruneRule.java:151]:
```suggestion
    private List<Long> rangePartitionPrune(OlapTable olapTable, RangePartitionInfo partitionInfo,
```

**Pull request closed without merging**